cmake/package_tooling.cmake

```diff
@@ -0,0 +1,118 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# The Zephyr package tooling script provides a generic script mode interface
+# to the Zephyr CMake package and module structure.
+#
+# The purpose of this script is to provide a single entry for running any Zephyr
+# build tool that is executed during CMake configure time without creating a
+# complete build system.
+#
+# It does so by loading the Zephyr CMake modules specified with the 'MODULES'
+# argument.
+#
+# As basic set of modules are always loaded to ensure identical behavior when
+# running tooling identical to Zephyr CMake configure time.
+#
+# This script executes the given module identical to Zephyr CMake configure time.
+# The application source directory must be specified using
+# '-S <path-to-sample>'
+#
+# The build directory will default to current working directory but can be
+# controlled with: '-B <path-to-build>'
+#
+# For example, if you were invoking CMake for 'hello_world' sample as:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world
+#
+# you can invoke only dts module (and dependencies) as:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# It is also possible to pass additional build settings.
+# If you invoke CMake for 'hello_world' as:
+#
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay
+#
+# you just add the same argument to tooling like:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# Note: the samples CMakeLists.txt file is not processed by package tooling, so
+#       any 'set(<var> <value>)' specified before 'find_package(Zephyr)' must be
+#       manually applied, for example if the CMakeLists.txt contains:
+#          set(OVERLAY_CONFIG foo.overlay)
+#          find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+#       the 'foo.overlay' must be specified using '-DOVERLAY_CONFIG=foo.overlay'
+
+cmake_minimum_required(VERSION 3.20)
+
+# add_custom_target and set_target_properties are not supported in script mode.
+# However, several Zephyr CMake modules create custom target for user convenience
+# like menuconfig, boards, shields, etc.
+# As we are not generating a build system with this tool, only running part of
+# the modules, then we simply override those functions to allow running those
+# modules.
+function(add_custom_target)
+  # This silence the error: 'add_custom_target command is not scriptable'
+endfunction()
+
+function(set_target_properties)
+  # This silence the error: 'set_target_properties command is not scriptable'
+endfunction()
```

Comment:
Why are we invoking CMake in script mode?

Reply:
because we don't want to start creating an actual build system with targets, etc.

Also, if this was a real project, then you would execute:
```
$ cmake -DBOARD=<board> -S <path-to>/package_tooling
```

so how will you specify the actual sample for which you want to run this tool ?
You cannot do:
```
$ cmake -DBOARD=<board> -S <path-to>/package_tooling -S samples/hello_world
```

I try to keep the invocation aligned so that `west build --some-new-arg` and `cmake ...` are closely aligned to creating the build system.

For example, generating build system:
```
$ cmake -DBOARD=<board> -B build -S samples/hello_world
``` 

Running the tool:
```
$ cmake -DBOARD=<board> -B build -S samples/hello_world -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
```

This means, everything before ` -DMODULES` is identical when creating a build system as well as running the tool.

Reply:
Understood.

I feel like this is a lot of code (which will need to be maintained as more functions that are not supported in script mode are added), just for the optimization of omitting the generation step.

But it doesn't increase the complexity of the normal build, so this is acceptable.

Reply:
"so how will you specify the actual sample for which you want to run this tool ?"

The alternative would be for the tool to not exist and for users to invoke CMake normally.